File: javascript/packages/linter/src/rules/html-allowed-script-type.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"
import { getTagLocalName, getAttribute, getStaticAttributeValue, hasAttributeValue } from "@herb-tools/core"
import { z } from "zod"

import type { BaseAutofixContext, UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLAttributeNode, HTMLOpenTagNode, ParseResult } from "@herb-tools/core"

export interface HTMLAllowedScriptTypeOptions {
  allowedTypes: string[]
  allowBlank: boolean
}

class AllowedScriptTypeVisitor extends BaseRuleVisitor {
  private readonly options: HTMLAllowedScriptTypeOptions

  constructor(ruleName: string, context: Partial<LintContext> | undefined, options: HTMLAllowedScriptTypeOptions) {
    super(ruleName, context)

    this.options = options
  }

  visitHTMLOpenTagNode(node: HTMLOpenTagNode): void {
    if (getTagLocalName(node) === "script") {
      this.visitScriptNode(node)
    }
  }

  private visitScriptNode(node: HTMLOpenTagNode): void {
    const { allowBlank } = this.options
    const typeAttribute = getAttribute(node, "type")

    if (!typeAttribute) {
      if (!allowBlank) {
        this.addOffense("`type` attribute required for `<script>` tag.", node.location)
      }

      return
    }

    if (!hasAttributeValue(typeAttribute)) {
      this.addOffense(
        "Avoid using an empty `type` attribute on the `<script>` tag. Either set a valid type or remove the attribute entirely.",
        typeAttribute.location
      )

      return
    }

    this.validateTypeAttribute(typeAttribute)
  }

  private validateTypeAttribute(typeAttribute: HTMLAttributeNode): void {
    const { allowedTypes, allowBlank } = this.options
    const typeValue = getStaticAttributeValue(typeAttribute)
    if (typeValue === null) return

    if (typeValue === "") {
      this.addOffense(
        "Avoid using an empty `type` attribute on the `<script>` tag. Either set a valid type or remove the attribute entirely.",
        typeAttribute.location
      )

      return
    }

    if (allowedTypes.includes(typeValue)) return

    this.addOffense(
      `Avoid using \`${typeValue}\` as the \`type\` attribute for the \`<script>\` tag. ` +
      `Must be one of: ${allowedTypes.map(t => `\`${t}\``).join(", ")}` +
      `${allowBlank ? " or blank" : ""}.`,
      typeAttribute.location
    )
  }
}

export class HTMLAllowedScriptTypeRule extends ParserRule<BaseAutofixContext, HTMLAllowedScriptTypeOptions> {
  static ruleName = "html-allowed-script-type"
  static introducedIn = this.version("0.9.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  get defaultOptions(): HTMLAllowedScriptTypeOptions {
    return {
      allowedTypes: ["text/javascript", "module", "importmap", "speculationrules", "application/ld+json"],
      allowBlank: true
    }
  }

  get optionsSchema(): z.ZodType<HTMLAllowedScriptTypeOptions> {
    return z.object({
      allowedTypes: z.array(z.string()),
      allowBlank: z.boolean()
    }).strict()
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new AllowedScriptTypeVisitor(this.ruleName, context, this.options)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/html_allowed_script_type.rs
use crate::utils::tag_utils::{get_attribute, get_static_attribute_value, get_tag_name_from_open_tag};

use herb::nodes::HTMLOpenTagNode;
use herb::Visitor;

const ALLOWED_TYPES: &[&str] = &["text/javascript"];
const ALLOW_BLANK: bool = true;

rule_visitor!(AllowedScriptTypeVisitor);
define_parser_rule!(HTMLAllowedScriptTypeRule, "html-allowed-script-type", Error, AllowedScriptTypeVisitor);

impl Visitor for AllowedScriptTypeVisitor {
  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    if let Some(tag_name) = get_tag_name_from_open_tag(node) {
      if tag_name.eq_ignore_ascii_case("script") {
        self.check_script_node(node);
      }
    }

    self.walk_html_open_tag_node(node);
  }
}

impl AllowedScriptTypeVisitor {
  fn check_script_node(&mut self, node: &HTMLOpenTagNode) {
    let type_attribute = match get_attribute(node, "type") {
      Some(attribute) => attribute,
      None => {
        if !ALLOW_BLANK {
          self.add_offense("`type` attribute required for `<script>` tag.".to_string(), node.location.clone());
        }

        return;
      }
    };

    if type_attribute.value.is_none() {
      self.add_offense(
        "Avoid using an empty `type` attribute on the `<script>` tag. Either set a valid type or remove the attribute entirely.".to_string(),
        type_attribute.location.clone(),
      );

      return;
    }

    let type_value = match get_static_attribute_value(type_attribute) {
      Some(value) => value,
      None => return,
    };

    if type_value.is_empty() {
      self.add_offense(
        "Avoid using an empty `type` attribute on the `<script>` tag. Either set a valid type or remove the attribute entirely.".to_string(),
        type_attribute.location.clone(),
      );

      return;
    }

    if ALLOWED_TYPES.contains(&type_value.as_str()) {
      return;
    }

    let allowed_list = ALLOWED_TYPES.iter().map(|type_name| format!("`{}`", type_name)).collect::<Vec<_>>().join(", ");

    let blank_suffix = if ALLOW_BLANK { " or blank" } else { "" };

    self.add_offense(
      format!(
        "Avoid using `{}` as the `type` attribute for the `<script>` tag. Must be one of: {}{}.",
        type_value, allowed_list, blank_suffix
      ),
      type_attribute.location.clone(),
    );
  }
}
